web/src/hooks/useSearch.ts

```diff
@@ -0,0 +1,223 @@
+        break;
+
+      case "done":
+        // add any potential followup questions
+        const followups = Object.entries(data)
+          .filter(([key]) => key.startsWith("followup_"))
+          .map(([k, value]) => value as Followup);
+        return { result, followups };
+      case "error":
+        throw data.error;
+    }
+  }
+  return { result, followups };
+};
+
+const fetchLLM = async (
+  query: string,
+  mode: string,
+  history: HistoryEntry[]
+): Promise<Response> =>
+  fetch(API_URL + "/chat", {
+    method: "POST",
+    cache: "no-cache",
+    keepalive: true,
+    headers: {
+      "Content-Type": "application/json",
+      Accept: "text/event-stream",
+      "Allow-Control-Allow-Origin": "*",
+    },
+
+    body: JSON.stringify({ query, mode, history }),
+  });
+
+export const queryLLM = async (
+  query: string,
+  mode: string,
+  history: HistoryEntry[],
+  baseReferencesIndex: number,
+  setCurrent: (e?: CurrentSearch) => void
+): Promise<SearchResult> => {
+  // do SSE on a POST request.
+  const res = await fetchLLM(query, mode, history);
+
+  if (!res.ok) {
+    return { result: { role: "error", content: "POST Error: " + res.status } };
+  }
+
+  try {
+    return await extractAnswer(res, baseReferencesIndex, setCurrent);
+  } catch (e) {
+    return {
+      result: { role: "error", content: e ? e.toString() : "unknown error" },
+    };
+  }
+};
+
+const cleanStampyContent = (contents: string) => contents.replace(
+    /<a(.*?)href="\/\?state=([a-zA-Z0-9]+.*?)"(.*?)<\/a>/g,
+    (_, pre, linkParts, post) => `<a${pre}href="${STAMPY_URL}/?state=${linkParts}"${post}</a>`
+);
+
+export const getStampyContent = async (
+  question_id: string
+): Promise<SearchResult> => {
+  const res = await fetch(`${STAMPY_CONTENT_URL}/${question_id}`, {
+    method: "GET",
+    headers: {
+      "Content-Type": "application/json",
+      Accept: "application/json",
+      "Allow-Control-Allow-Origin": "*",
+    },
+  });
+
+  if (!res.ok) {
+    return { result: { role: "error", content: "POST Error: " + res.status } };
+  }
+
+  const data = (await res.json()).data;
+
+  let result = {
+    role: "stampy",
+    content: cleanStampyContent(data.text),
+    url: `${STAMPY_URL}/?state=${data.pageid}`,
+  } as StampyMessage;
+
+  // re-enable the searchbox, with the question that was just answered
+  // removed from the list of possible followups.
+
+  // create an array of new followup questions from the data
+  const f_new = data.relatedQuestions.map((f: any) => ({
+    pageid: f.pageid!,
+    text: f.title!,
+    score: 0,
+  }));
+
+  const fpids = new Set(f_new.map((f: Followup) => f.pageid));
+  const followups = (f_old: Followup[]): Followup[] => {
+    const f_old_filtered = f_old.filter(
+      (f) => f.pageid !== data.pageid && !fpids.has(f.pageid)
+    );
+    return [...f_new, ...f_old_filtered].slice(0, MAX_FOLLOWUPS); // this is correct, it's N and not N-1 in javascript fsr
```

Comment:
not sure what this commen this pointing at, does a slice work differently in some language?
Python slice to pick first 3 items like `[0, 1, 2][:3]` works the same as JS

...and `range(3)` in Python is also the same as `Array.from(Array(3)).map((_, i) => i)` in JS 🤷 

also, what is `fsr` please?

Reply:
dunno. This I just moved over. Will remove